test/distributed/_composable/fsdp/test_fully_shard_comm.py

```diff
@@ -94,5 +104,240 @@ def reduce_scatter_with_count(*args, **kwargs):
         self.assertEqual(reduce_scatter_count, num_blocks + 1)
 
 
+class TestFullyShardBackwardPrefetch(FSDPTest):
+    @property
+    def world_size(self) -> int:
+        return min(4, torch.cuda.device_count())
+
+    @skip_if_lt_x_gpu(2)
+    def test_fully_shard_backward_prefetch(self):
+        self.run_subtests(
+            {"reshard_after_forward": [True, False, 2]},
+            self._test_backward_prefetch_forward_backward,
+        )
+        self.run_subtests(
+            {"reshard_after_forward": [True, False, 2]},
+            self._test_backward_prefetch_multi_forward,
+        )
+        self._test_backward_prefetch_unused_in_backward(True)
+
+    def _test_backward_prefetch_forward_backward(
+        self, reshard_after_forward: Union[bool, int]
+    ):
+        n_layers = 3
+        model, optim, inp = self._init_transformer(n_layers, reshard_after_forward)
+        events: List[EventType] = []
+        unshard_with_record = self._get_unshard_with_record(
+            FSDPParamGroup.unshard, events
+        )
+        post_backward_with_record = self._get_post_backward_with_record(
+            FSDPParamGroup.post_backward, events
+        )
+        # Check the order for normal 1 forward, 1 backward, 1 optimizer step
+        with patch_unshard(unshard_with_record), patch_post_backward(
```

Comment:
Hmmm, it's a bit hard to follow the user workflow when using reshard_after_forward and backward prefetch. and the reshard_after_forward feature seems getting FSDP more complicated than my original thought..

Reply:
Oh I see, I thought there needs some special API for user to derive the backward order events for backward prefetching in the case of reshard_after_forward, if this happens automatically then it should be fine.